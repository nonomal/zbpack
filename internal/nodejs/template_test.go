package nodejs_test

import (
	"os"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"github.com/zeabur/zbpack/internal/nodejs"
)

func TestMain(m *testing.M) {
	v := m.Run()

	// After all tests have run `go-snaps` will sort snapshots
	snaps.Clean(m, snaps.CleanOpts{Sort: true})

	os.Exit(v)
}

func TestTemplate_NBuildCmd_NOutputDir(t *testing.T) {
	ctx := nodejs.TemplateContext{
		NodeVersion: "18",

		InstallCmd: "RUN yarn install",
		BuildCmd:   "",
		StartCmd:   "yarn start",
	}

	result, err := ctx.Execute()
	assert.NoError(t, err)
	snaps.MatchSnapshot(t, result)
}

func TestTemplate_NBuildCmd_OutputDir_NSPA(t *testing.T) {
	ctx := nodejs.TemplateContext{
		NodeVersion: "18",

		InstallCmd: "RUN yarn install",
		BuildCmd:   "",
		StartCmd:   "yarn start",
	}

	result, err := ctx.Execute()
	assert.NoError(t, err)
	snaps.MatchSnapshot(t, result)
}

func TestTemplate_NBuildCmd_OutputDir_SPA(t *testing.T) {
	ctx := nodejs.TemplateContext{
		NodeVersion: "18",

		InstallCmd: "RUN yarn install",
		BuildCmd:   "",
		StartCmd:   "yarn start",
	}

	result, err := ctx.Execute()
	assert.NoError(t, err)
	snaps.MatchSnapshot(t, result)
}

func TestTemplate_BuildCmd_NOutputDir(t *testing.T) {
	ctx := nodejs.TemplateContext{
		NodeVersion: "18",

		InstallCmd: "RUN yarn install",
		BuildCmd:   "yarn build",
		StartCmd:   "yarn start",

		Serverless: true,
	}

	result, err := ctx.Execute()
	assert.NoError(t, err)
	snaps.MatchSnapshot(t, result)
}

func TestTemplate_BuildCmd_OutputDir(t *testing.T) {
	ctx := nodejs.TemplateContext{
		NodeVersion: "18",

		InstallCmd: "RUN yarn install",
		BuildCmd:   "yarn build",
		StartCmd:   "yarn start",

		OutputDir: "/app/dist",
	}

	result, err := ctx.Execute()
	assert.NoError(t, err)
	snaps.MatchSnapshot(t, result)
}

func TestTemplate_BuildCmd_Bun(t *testing.T) {
	ctx := nodejs.TemplateContext{
		Bun:         true,
		BunVersion:  "1.2",
		NodeVersion: "18",
		InstallCmd:  "RUN bun install",
		StartCmd:    "bun start main.ts",
	}

	result, err := ctx.Execute()
	assert.NoError(t, err)
	snaps.MatchSnapshot(t, result)
}

func TestTemplate_Monorepo(t *testing.T) {
	ctx := nodejs.TemplateContext{
		NodeVersion: "18",
		AppDir:      "myservice",
		InstallCmd:  "WORKDIR /src/myservice\nRUN yarn install",
		StartCmd:    "yarn start",
	}

	result, err := ctx.Execute()
	assert.NoError(t, err)
	snaps.MatchSnapshot(t, result)
}

func TestTemplate_MonorepoServerless(t *testing.T) {
	ctx := nodejs.TemplateContext{
		NodeVersion: "18",
		AppDir:      "myservice",
		InstallCmd:  "WORKDIR /src/myservice\nRUN yarn install",
		StartCmd:    "yarn start",
		Serverless:  true,
	}

	result, err := ctx.Execute()
	assert.NoError(t, err)
	snaps.MatchSnapshot(t, result)
}

func TestTemplate_MonorepoServerlessOutDir(t *testing.T) {
	ctx := nodejs.TemplateContext{
		NodeVersion: "18",
		AppDir:      "myservice",
		InstallCmd:  "WORKDIR /src/myservice\nRUN yarn install",
		StartCmd:    "yarn start",
		OutputDir:   "/app/dist",
	}

	result, err := ctx.Execute()
	assert.NoError(t, err)
	snaps.MatchSnapshot(t, result)
}

func TestTemplate_Bun(t *testing.T) {
	ctx := nodejs.TemplateContext{
		NodeVersion: "18",
		InstallCmd:  "RUN bun install",
		StartCmd:    "bun start main.ts",
		Bun:         true,
		BunVersion:  "1",
	}

	result, err := ctx.Execute()
	assert.NoError(t, err)
	snaps.MatchSnapshot(t, result)
}

func TestTemplate_ServerlessOutputDir(t *testing.T) {
	t.Parallel()

	t.Run("with serverless", func(t *testing.T) {
		ctx := nodejs.TemplateContext{
			NodeVersion: "18",
			InstallCmd:  "RUN yarn install",
			Serverless:  true,
			OutputDir:   "/app/dist",
		}

		result, err := ctx.Execute()
		assert.NoError(t, err)

		require.Contains(t, result, "FROM scratch AS output")
		require.NotContains(t, result, "FROM zeabur/caddy-static AS runtime")
	})

	t.Run("without serverless", func(t *testing.T) {
		ctx := nodejs.TemplateContext{
			NodeVersion: "18",
			InstallCmd:  "RUN yarn install",
			Serverless:  false,
			OutputDir:   "/app/dist",
		}

		result, err := ctx.Execute()
		assert.NoError(t, err)

		require.Contains(t, result, "FROM scratch AS output")
		require.Contains(t, result, "FROM zeabur/caddy-static AS runtime")
	})
}
