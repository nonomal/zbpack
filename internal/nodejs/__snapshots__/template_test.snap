
[TestTemplate_BuildCmd_Bun - 1]
# Install bun if we need it
FROM oven/bun:1.2 AS bun-runtime
FROM node:18 AS build

ENV PORT=8080
WORKDIR /src
# Copy the bun binary from the bun-runtime stage directly.
# A bit hacky but it works.
COPY --from=bun-runtime /usr/local/bin/bun /usr/local/bin
COPY --from=bun-runtime /usr/local/bin/bunx /usr/local/bin

# Check if we have 'corepack' available; if none, we
# install corepack@0.10.0.
RUN which corepack || npm install -g --force corepack@0.10.0
RUN corepack enable

RUN bun install

COPY . .
# Build if we can build it


EXPOSE 8080
CMD bun start main.ts

---

[TestTemplate_BuildCmd_NOutputDir - 1]
FROM node:18 AS build

ENV PORT=8080
WORKDIR /src

# Check if we have 'corepack' available; if none, we
# install corepack@0.10.0.
RUN which corepack || npm install -g --force corepack@0.10.0
RUN corepack enable

RUN yarn install

COPY . .
# Build if we can build it
RUN yarn build

FROM scratch AS output
COPY --from=build /src/ /


---

[TestTemplate_BuildCmd_OutputDir - 1]
FROM node:18 AS build

ENV PORT=8080
WORKDIR /src

# Check if we have 'corepack' available; if none, we
# install corepack@0.10.0.
RUN which corepack || npm install -g --force corepack@0.10.0
RUN corepack enable

RUN yarn install

COPY . .
# Build if we can build it
RUN yarn build

FROM scratch AS output
COPY --from=build /src///app/dist /

FROM zeabur/caddy-static AS runtime
COPY --from=output / /usr/share/caddy



---

[TestTemplate_Bun - 1]
# Install bun if we need it
FROM oven/bun:1 AS bun-runtime
FROM node:18 AS build

ENV PORT=8080
WORKDIR /src
# Copy the bun binary from the bun-runtime stage directly.
# A bit hacky but it works.
COPY --from=bun-runtime /usr/local/bin/bun /usr/local/bin
COPY --from=bun-runtime /usr/local/bin/bunx /usr/local/bin

# Check if we have 'corepack' available; if none, we
# install corepack@0.10.0.
RUN which corepack || npm install -g --force corepack@0.10.0
RUN corepack enable

RUN bun install

COPY . .
# Build if we can build it


EXPOSE 8080
CMD bun start main.ts

---

[TestTemplate_Monorepo - 1]
FROM node:18 AS build

ENV PORT=8080
WORKDIR /src

# Check if we have 'corepack' available; if none, we
# install corepack@0.10.0.
RUN which corepack || npm install -g --force corepack@0.10.0
RUN corepack enable

WORKDIR /src/myservice
RUN yarn install


# Build if we can build it


EXPOSE 8080
CMD yarn start

---

[TestTemplate_MonorepoServerless - 1]
FROM node:18 AS build

ENV PORT=8080
WORKDIR /src

# Check if we have 'corepack' available; if none, we
# install corepack@0.10.0.
RUN which corepack || npm install -g --force corepack@0.10.0
RUN corepack enable

WORKDIR /src/myservice
RUN yarn install


# Build if we can build it


FROM scratch AS output
COPY --from=build /src/myservice /


---

[TestTemplate_MonorepoServerlessOutDir - 1]
FROM node:18 AS build

ENV PORT=8080
WORKDIR /src

# Check if we have 'corepack' available; if none, we
# install corepack@0.10.0.
RUN which corepack || npm install -g --force corepack@0.10.0
RUN corepack enable

WORKDIR /src/myservice
RUN yarn install


# Build if we can build it


FROM scratch AS output
COPY --from=build /src/myservice//app/dist /

FROM zeabur/caddy-static AS runtime
COPY --from=output / /usr/share/caddy



---

[TestTemplate_NBuildCmd_NOutputDir - 1]
FROM node:18 AS build

ENV PORT=8080
WORKDIR /src

# Check if we have 'corepack' available; if none, we
# install corepack@0.10.0.
RUN which corepack || npm install -g --force corepack@0.10.0
RUN corepack enable

RUN yarn install

COPY . .
# Build if we can build it


EXPOSE 8080
CMD yarn start

---

[TestTemplate_NBuildCmd_OutputDir_NSPA - 1]
FROM node:18 AS build

ENV PORT=8080
WORKDIR /src

# Check if we have 'corepack' available; if none, we
# install corepack@0.10.0.
RUN which corepack || npm install -g --force corepack@0.10.0
RUN corepack enable

RUN yarn install

COPY . .
# Build if we can build it


EXPOSE 8080
CMD yarn start

---

[TestTemplate_NBuildCmd_OutputDir_SPA - 1]
FROM node:18 AS build

ENV PORT=8080
WORKDIR /src

# Check if we have 'corepack' available; if none, we
# install corepack@0.10.0.
RUN which corepack || npm install -g --force corepack@0.10.0
RUN corepack enable

RUN yarn install

COPY . .
# Build if we can build it


EXPOSE 8080
CMD yarn start

---
